mm2src/common/custom_futures/repeatable.rs

```diff
@@ -0,0 +1,785 @@
+#[derive(Debug)]
+pub enum Action<T, E> {
+    Ready(T),
+    Retry(E),
+}
+
+pub trait RetryOnError<T, E> {
+    fn retry_on_err(self) -> Action<T, E>;
+}
+
+impl<T, E> RetryOnError<T, E> for Result<T, E> {
+    /// Converts `Result<T, E>` into `Action<T, E>`:
+    /// * `Ok(T)` => `Action::Ready(T)`.
+    /// * `Err(E)` => `Action::Retry(E)`.
+    #[inline]
+    fn retry_on_err(self) -> Action<T, E> {
+        match self {
+            Ok(ready) => Action::Ready(ready),
+            Err(e) => Action::Retry(e),
+        }
+    }
+}
+
+/// The result of `repeatable` or `retry_on_err` macros - the first step at the future configuration.
+pub struct Repeatable<Factory, F, T, E> {
+    factory: Factory,
+    /// Currently executable future, i.e. an active attempt.
+    exec_fut: Option<F>,
+    /// A timeout future if we're currently waiting for a timeout.
+    timeout_fut: Option<Timer>,
+    until: Option<Until>,
+    attempts: Option<AttemptsState>,
+    repeat_every: Option<Duration>,
+    inspect_err: Option<Box<dyn InspectErrorTrait<E>>>,
+    _phantom: PhantomData<(F, T, E)>,
+}
+
+impl<Factory, F, T, E> Repeatable<Factory, F, T, E>
+where
+    Factory: FactoryTrait<F>,
+    F: RepeatableTrait<T, E>,
+{
+    #[inline]
+    pub fn new(factory: Factory) -> Self {
+        let exec = factory();
+
+        Repeatable {
+            factory,
+            exec_fut: Some(exec),
+            timeout_fut: None,
+            until: None,
+            attempts: None,
+            repeat_every: None,
+            inspect_err: None,
+            _phantom: PhantomData::default(),
+        }
+    }
+
+    /// Specifies an inspect handler that does something with an error on each unsuccessful attempt.
+    #[inline]
+    pub fn inspect_err<Inspect>(mut self, inspect: Inspect) -> Self
+    where
+        Inspect: 'static + Fn(&E) + Send,
+    {
+        self.inspect_err = Some(Box::new(inspect));
+        self
+    }
+
+    #[inline]
+    pub fn repeat_every(mut self, repeat_every: Duration) -> Self {
+        self.repeat_every = Some(repeat_every);
+        self
+    }
+
+    #[inline]
+    pub fn repeat_every_ms(self, repeat_every: u64) -> Self { self.repeat_every(Duration::from_millis(repeat_every)) }
+
+    #[inline]
+    pub fn repeat_every_secs(self, repeat_every: f64) -> Self {
+        self.repeat_every(Duration::from_secs_f64(repeat_every))
+    }
+
+    /// Specifies a total number of attempts to run the future.
+    /// So there will be up to `total_attempts`.
+    #[inline]
+    pub fn attempts(mut self, total_attempts: usize) -> Self {
+        if total_attempts == 0 {
+            warn!("There will be 1 attempt even though 'total_attempts' is 0");
+        }
+
+        self.attempts = Some(AttemptsState {
+            current_attempt: 0,
+            total_attempts,
+        });
+        self
+    }
+
+    /// Specifies a deadline before that we may try to repeat the future.
+    #[inline]
+    pub fn until(mut self, until: Instant) -> Self {
```

Comment:
This is used only in tests. Is it really required?

Reply:
When I added this, I thought that it will be used. Now I see that it's not currently used, and also requires to add `Until` enum that can be easily replaced.
Will remove this

Reply:
Removed